### PR TITLE
Feature/full code cleanup

### PR DESCRIPTION
**Pull request by have-fun-was-taken**
One by one, I muted the edits that a full Code Cleanup of Visual Studio would do, until the cleanup did no changes. Then I unmuted them one-per-commit so they are readable.  
This PR also contains a couple of manual edits and refactors. It is indicated in the commit message when the changes are manual. Feel free to ask questions if there are any.

Regarding the [comment](https://github.com/have-fun-was-taken/yafc-ce/pull/58#discussion_r1501895474) of @SWeini about `var`, this time I set `csharp_style_var_elsewhere = false:none`, so the previous example of `foreach (var analysis in analyses)` was kept with `var`.

QA: loaded a usual save, checked that the basic functionality works.

### CODE REVIEW

File: YAFCmodel/Data/DataUtils.cs
Comment (veger) [inline YAFCmodel/Data/DataUtils.cs]: Personally, I am not a big fan of this, I do not find it clear(er), but I can adjust


IDE will help me getting this syntax right :smile:

File: .editorconfig
Comment (veger) [inline .editorconfig]: Even though this explicitly shows the return(ed) value is ignored, I doubt it will add any value (like preventing a but due to ignored value)

Note that in golang this is even required by the compiler, so it is a (forced) common practice for me. I just have my doubts for the usefulness in C#/this project.

File: YAFC/Widgets/DataGrid.cs
Comment (veger) [inline YAFC/Widgets/DataGrid.cs:65]: here we are putting the parenthesis back (see one of my previous comments were  the unnecessary ones were removed). Is sounds conflicting?

File: .editorconfig
Comment (veger) [inline .editorconfig]: You make it a suggestion, but fix all/a lot? Sounds like it is mandatory :smile: 
(I am fine either way)

**Review by veger — COMMENTED** [review]
Nice work by splitting them into separate commits: easy to review this huge PR!

File: YAFCui/ImGui/ImGuiUtils.cs
Comment (have-fun-was-taken) [reply YAFCui/ImGui/ImGuiUtils.cs:414]: This one is https://github.com/have-fun-was-taken/yafc-ce/pull/64/commits/559847e35741e11a105a5f864654bb9eb9693125
>  Manual refactor: make SearchQuery readonly

> Due to how SetSearch was implemented SearchQuery could not be made
readonly. However, Visual Studio would occasionally bug out and
make it readonly when running Code Cleanup, so that raised a choice:
whether to mute the inspection or to refactor the class.

> I refactored the class by removing SetSearch and making SearchQuery
readonly.

File: .editorconfig
Comment (shihan42) [reply .editorconfig]: Depends. If we add some whitespaces for alignment, it suddenly becomes very readable :)


```csharp
// -> YAFC/YafcLib.cs @ line 32
return libraryname switch {
    "lua52" => "liblua52.so",
    "SDL2.dll" => "SDL2-2.0.so.0",
    "SDL2_ttf.dll" => "SDL2_ttf-2.0.so.0",
    "SDL2_image.dll" => "SDL2_image-2.0.so.0",
    _ => libraryname,
};

// -> 

return libraryname switch
{
    "lua52"          => "liblua52.so",
    "SDL2.dll"       => "SDL2-2.0.so.0",
    "SDL2_ttf.dll"   => "SDL2_ttf-2.0.so.0",
    "SDL2_image.dll" => "SDL2_image-2.0.so.0",
    _                => libraryname,
};


// -> YAFCmodel/Model/Project.cs @ line 180

return time switch {
    1 or 0 => (1f, "s"),
    60 => (1f / 60f, "m"),
    3600 => (1f / 3600f, "h"),
    _ => (1f / time, "t"),
};

// ->

return time switch
{
    1 or 0 => (1f, "s"),
    60     => (1f / 60f, "m"),
    3600   => (1f / 3600f, "h"),
    _      => (1f / time, "t"),
};
```

File: .editorconfig
Comment (shihan42) [reply .editorconfig]: It removes the unnecessary clutter of == null
But I admit, this is more or less a matter of habit.

File: .editorconfig
Comment (shihan42) [reply .editorconfig]: Regarding this project, I'm also unsure.
But generally speaking, it helps refactoring. Just search for "_ =" and mark all these methods for refactoring. If their return value is not used, why return it in the first place?

File: .editorconfig
Comment (shihan42) [reply .editorconfig]: I'd say make it mandatory. Suggestions are neither this nor that. Make it clear! :)

File: YAFC/Widgets/DataGrid.cs
Comment (shihan42) [reply YAFC/Widgets/DataGrid.cs:65]: I think this one should be preferred, but is just my personal opinion. My head is faster detecting the the parentheses than parsing operator precedence ;)

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: If we look at the actual changes done by this commit, an argument can be made that the double equality in the [ternary](https://github.com/have-fun-was-taken/yafc-ce/pull/64/commits/9c94f5848834eaa11e554365927ff7fa617a0e94#diff-83fecb21d3d33f0ef75e66c786eab4e5cd0c8e4e309117e3b75a6be78649a91bR221) is less understandable, but barely. Do you not consider the rest of the changes in this commit an improvement?

File: YAFCmodel/Data/DataUtils.cs
Comment (have-fun-was-taken) [reply YAFCmodel/Data/DataUtils.cs]: Agreed. I can justify this pattern matching [here](https://github.com/have-fun-was-taken/yafc-ce/pull/64/files/fe429f64dba69c1c8fd46a1c064c418b81dc71c9..7e869750f5da2a87812a6af8e89fd91d012ba250#diff-574901d659ae628ac0805d2c373c7a94d404780ed54a5e7d92988e8c4ad272bbR281), but not at the line you mentioned.

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: > why return it in the first place?

Because they are external-library methods. It's especially prevalent in the parts of code that use the SDL library. I'll take another look at the result to see if it really improves readability.

File: YAFC/Widgets/DataGrid.cs
Comment (have-fun-was-taken) [reply YAFC/Widgets/DataGrid.cs:65]: > sounds conflicting?

The rules were unmuted one by one. Some of them overwrite each other, so it indeed could happen that some parentheses were put back.

File: .editorconfig
Comment (shihan42) [reply .editorconfig]: Ah, ok. I didn't mean to get into specifics there, I was just speaking more generally.
3rd party methods are a different thing, yes.

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: > You make it a suggestion, but fix all

The Code Cleanup applies the fix if the rule is `suggestion` or higher. Not sure what would happen if there were a [`silent`](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-options#severity-level) rule, but Code Cleanup doesn't apply fixes if the severity is `none`.

In other words, the severity `suggestion` means that Code Cleanup applies the fix, and the rule is non-mandatory for the dev.

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: The point is that the most Discards are indeed used for the library method. That raises the question whether we need the Discards at all as we cannot fix the issue because they are 3rd-party methods. I'll think on that.

**Review by shihan42 — APPROVED** [review]
If @SWeini doesn't object, I think it's ready to be merged.

**Comment by have-fun-was-taken** [comment]
In the meantime, I'll adress the concerns that @veger pointed out. 

On a side note, please _don't_ merge PRs into master even if you have the ability to do so. Otherwise, you could merge something that needs fixes.

**Comment by shihan42** [comment]
> On a side note, please _don't_ merge PRs into master even if you have the ability to do so. Otherwise, you could merge something that needs fixes. I was procrastinating on making a proper permission system, but I'll need to get it done eventually.

Do you mean until this PR is done? Or generally speaking?

**Comment by have-fun-was-taken** [comment]
Generally speaking.

**Comment by have-fun-was-taken** [comment]
I'll consult with my former colleagues on the exact reasoning behind it to present a better argument about it.

**Comment by SWeini** [comment]
> If @SWeini doesn't object, I think it's ready to be merged.

no objections without looking too closely
I don't care about coding style as long as its consistent and the tooling is configured to allow auto-format

**Comment by have-fun-was-taken** [comment]
> I'll consult with my former colleagues on the exact reasoning behind it to present a better argument about it.

After checking with the GTNH community, it seems that I misremembered. Anyone with the permissions can merge.
Regarding this PR in particular, I ask you to let me merge because I need to think on what @veger mentioned.

File: YAFCui/ImGui/ImGuiUtils.cs
Comment (veger) [reply YAFCui/ImGui/ImGuiUtils.cs:414]: Yes I saw, I reviewed the separate commits. I was just wondering about the impact when the user is doing some searches.

Comment (have-fun-was-taken) [reply YAFCui/ImGui/ImGuiUtils.cs:414]: The previous implementation essentially did the [same](https://github.com/have-fun-was-taken/yafc-ce/commit/559847e35741e11a105a5f864654bb9eb9693125#diff-aab5f4c0184c7f6e8dcafd6e14852be6ba530f125c9fb464c146501dcb7ebcaaL10) under another function, so there should be no visible changes to the performance.

File: .editorconfig
Comment (veger) [reply .editorconfig]: I was only talking about these kind of constractions:
```
rect.Width *= (1f - waste);
```

I know that lots of devs that are unsure about how this works without the parenthesis. And with them, there is suddenly no issue anymore. As said either way I am fine as I do know they are the same :smile: It was just as we discussed to make on boarding new contributes as easy as possible (not to scare them away) that I wondered about this change.

Other changes like:
```
return ((moduleCount - 1) / beacon.moduleSlots + 1);
```
to
```
return (moduleCount - 1) / beacon.moduleSlots + 1;
```
I totally agree on, as there absolutely no benefit of having the extra parenthesis here.

File: YAFCui/ImGui/ImGuiUtils.cs
Comment (veger) [reply YAFCui/ImGui/ImGuiUtils.cs:414]: In the previous implementation the `SearchQuery` object was reused (for some reason). With the proposed one the `SearchQuery` is thrown away and a new one created (each time you type a new char in the search box?)

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: I'll see if there's a separate setting for the class of operators you mentioned.

File: .editorconfig
Comment (veger) [reply .editorconfig]: I do not have _any_ habit for C#, so as with most things either way is fine for me personally as long as it helps the broader/potential audience

File: YAFCmodel/Data/DataUtils.cs
Comment (veger) [reply YAFCmodel/Data/DataUtils.cs]: Yes those lines are very nice and readable with this change. For them I really do like it a lot.

File: YAFCui/ImGui/ImGuiUtils.cs
Comment (have-fun-was-taken) [reply YAFCui/ImGui/ImGuiUtils.cs:414]: The previous implementation had the following line:
```
// YAFCui/Core/SearchQuery.cs
public void SetSearch(string query) {
    this = new SearchQuery(query);
}
```
Based on this line, I assume that the `SearchQuery` object was not reused because `this` was assigned to a new object. Is my assumption incorrect?

**Comment by veger** [comment]
With the branch protection (of the `master` branch), you can force a PR, so no direct merging. You can also require PR to have a certain amount of approvals. With this you force reviews before merging to the branch.

If this is 'too much' for you, you can exempt administrators from these rules (so you can still merge to master without approvals/PRs :wink:  )

**Comment by veger** [comment]
This are our (partial) branch protection rules we use to enforce PRs/reviews:
image

and

image

**Comment by have-fun-was-taken** [comment]
> With the branch protection (of the master branch), you can force using PRs

This is how it is set up right now. The `master` branch is PR-protected with the requirement of one review. I think this pipeline is fitting for the current scale of the project.

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: > `// -> YAFC/YafcLib.cs @ line 32`

Did you make these references automatically somehow? If yes, I'd like to know how to do it too :D

Comment (have-fun-was-taken) [reply .editorconfig]: Regarding the formatting that @shihan42 suggested, it triggers [IDE0055](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0055) -- an all-encompassing formatting inspection. I haven't found a way to allow many spaces in switches yet.

Comment (veger) [reply .editorconfig]: Just leave it (the formatting without the alignment, formatting tools usually do not like these alignments...).

The example that @shihan42 showed indeed makes it more compact and faster to comprehend. So I think this is a nice change (in at least these/most cases). So lets mark this as resolved and keep the commit.

Comment (have-fun-was-taken) [reply .editorconfig]: Turns out it's a very long-lasting [issue](https://github.com/dotnet/roslyn/issues/38241) with no pretty solution.

Comment (veger) [reply .editorconfig]: 2019? that is indeed a long time... :disappointed:

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: Addressed in https://github.com/have-fun-was-taken/yafc-ce/pull/64/commits/614d701c4f075def24570b1f4aafff5a52817d83

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: > Is this default C# notation that is used by the broad(er) audience?

Unmuting this severity is essentially coming back to the Microsoft code style that we copied at the start, so I think it is safe to assume that it is a widespread convention.

File: YAFCmodel/Data/DataUtils.cs
Comment (have-fun-was-taken) [reply YAFCmodel/Data/DataUtils.cs]: The problem is that I don't see a setting to turn it off only for intense usages, and the used operators are largely the same. With parentheses enabled, it's sort-of readable: 
```
if (c is (>= '0' and <= '9') or '.' or '-' or 'e')
```
Therefore, I'm leaning towards leaving IDE0078 enabled because other use cases improve readability.

File: .editorconfig
Comment (have-fun-was-taken) [reply .editorconfig]: > The point is that the most Discards are indeed used for the library methods

After looking at the code more, I'm not that sure if my previous statement is true. There are many usages of the YAFC methods that simply don't use the return values. 

> Even though this explicitly shows the return(ed) value is ignored, I doubt it will add any value (like preventing a bug due to ignored value)

If not by preventing bugs, it will add value by reducing the amount of effort required to read the code. It will indicate to a reader that it's not a bug that YAFC uses its own functions without using the returned values.

**Comment by have-fun-was-taken** [comment]
@veger I left unresolved only the conversations that may need your further approval. Please resolve them when you consider them addressed to a satisfying conclusion.

File: YAFCui/ImGui/ImGuiUtils.cs
Comment (veger) [reply YAFCui/ImGui/ImGuiUtils.cs:414]: hm... I am not used to a language allowing to change `this`,`self`, etc... (it is kind of weird if you think about it)
But I guess it indeed creates a new object at the least (discarding the existing one)

File: YAFCmodel/Data/DataUtils.cs
Comment (veger) [reply YAFCmodel/Data/DataUtils.cs]: agreed, this is probably an outlier

File: .editorconfig
Comment (veger) [reply .editorconfig]: So they could be indicators for (outdated) code that needs some fixing? I can live with that :smile:

**Review by veger — APPROVED** [review]

**Comment by veger** [comment]
All of my comments are resolved. If the others also agree we can merge this

**Pull request merged**
